core/parachain/validator/parachain_processor.hpp

```diff
@@ -7,39 +7,294 @@
+    std::optional<network::ValidatorIndex> getOurIndex();
+
+    /*
+     * Helpers.
+     */
+    primitives::BlockHash candidateHashFrom(
+        FetchedCollationState const &fetched_collation_state) {
+      return visit_in_place(
+          fetched_collation_state.fetched_collation.response_data,
+          [&](network::CollationResponse const &collation_response)
+              -> primitives::BlockHash {
+            return hasher_->blake2b_256(
+                scale::encode(collation_response.receipt).value());
+          });
+    }
+
+    network::CandidateDescriptor const &candidateDescriptorFrom(
+        FetchedCollationState const &fetched_collation_state) {
+      return visit_in_place(
+          fetched_collation_state.fetched_collation.response_data,
+          [](network::CollationResponse const &collation_response)
+              -> network::CandidateDescriptor const & {
+            return collation_response.receipt.descriptor;
+          });
+    }
+
+    std::optional<std::reference_wrapper<network::CandidateDescriptor const>>
+    candidateDescriptorFrom(
+        std::shared_ptr<network::Statement> const &statement) {
+      return visit_in_place(
+          statement->candidate_state,
+          [](network::CommittedCandidateReceipt const &receipt)
+              -> std::optional<
+                  std::reference_wrapper<network::CandidateDescriptor const>> {
+            return receipt.descriptor;
+          },
+          [](...)
+              -> std::optional<
+                  std::reference_wrapper<network::CandidateDescriptor const>> {
+            BOOST_ASSERT(false);
+            return std::nullopt;
+          });
+    }
+
+    network::CollatorPublicKey const &collatorIdFromDescriptor(
+        network::CandidateDescriptor const &descriptor) {
+      return descriptor.collator_id;
+    }
+
+    primitives::BlockHash candidateHashFrom(
+        std::shared_ptr<network::Statement> const &statement) {
+      return visit_in_place(
+          statement->candidate_state,
+          [](network::Dummy const &) {
+            BOOST_ASSERT(!"Not used!");
+            return primitives::BlockHash{};
+          },
+          [&](network::CommittedCandidateReceipt const &data) {
+            return hasher_->blake2b_256(
+                scale::encode(network::CandidateReceipt{
+                                  .descriptor = data.descriptor,
+                                  .commitments_hash = hasher_->blake2b_256(
+                                      scale::encode(data.commitments).value())})
+                    .value());
+          },
+          [&](primitives::BlockHash const &candidate_hash) {
+            return candidate_hash;
+          });
+    }
+
+    /*
+     * Notification
+     */
+    template <typename F>
+    void notify_internal(std::shared_ptr<WorkersContext> &context, F &&func) {
+      BOOST_ASSERT(context);
+      boost::asio::post(*context, std::forward<F>(func));
+    }
```

Comment:
Do we need this wrapper?  
Better to find/use `NonNull` class, which will assert on `operator*` and `operator->`.